### PR TITLE
feat: Adding new samples for the python callouts

### LINKED ISSUE #64: Check client request for a bad token referencing a list in configuration data
## Use-case

1. Check client request for a bad token referencing a list in configuration data

## Details

UC8. Reference a list of bad tokens in plugin configuration data to check if a given request should be authorized or not. Log the authorization decision made by the plugin.

### PR DESCRIPTION
**Pull request by pweiber**
Fixes #64 considering #65 PR comments

## Use-cases

9. Validate client JWT for authorization
~~10. Validate client token using HMAC cookieless~~

## Details

UC9 - Check the client request (query string) for a valid JWT token, validate (using RS256) log the authorization decision in Cloud Logging. 

~~UC10 - Check the client request URL for a valid JWT token signed using HMAC. Log the authorization decision.~~

### CODE REVIEW

File: callouts/python/extproc/example/jwt_auth/service_callout_example.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+from extproc.service import callout_tools
+
+def check_request_header(request_headers, header_key):
+  """Check header of the request."""
+  return next((header.key
+                   for header in request_headers.headers.headers
+                   if header.key == header_key), None)
+
+def extract_jwt_token(context):
+  metadata = dict(context.invocation_metadata())
```
Comment (rshriram) [inline callouts/python/extproc/example/jwt_auth/service_callout_example.py:31]: isnt the JWT token part of some HTTP header like "Authorization: Bearer <...>" ?

File: callouts/python/extproc/example/jwt_auth/service_callout_example.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  jwt_token = metadata.get('authorization')
```
Comment (rshriram) [inline callouts/python/extproc/example/jwt_auth/service_callout_example.py:32]: isnt this coming from the client's HTTP headers?

File: callouts/python/extproc/example/jwt_auth/service_callout_example.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  extracted_jwt = jwt_token.split(' ')[1] if jwt_token and ' ' in jwt_token else jwt_token
+  return extracted_jwt
+
+def validate_jwt_authorization(key, context, algorithm):
```
Comment (rshriram) [inline callouts/python/extproc/example/jwt_auth/service_callout_example.py:36]: validate_jwt_token

File: callouts/python/extproc/example/jwt_auth/service_callout_example.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsImlhdCI6MTcxMjM1MTk5NywiZXhwIjoyMDc1MjMxOTk3fQ.sheYamXYaJsqRhTv1g_tyTbP9D7Q8Q3ANh2AgH7l46Q
+
+  """
+
+  def on_request_headers(
+      self, headers: service_pb2.HttpHeaders,
+      context: ServicerContext):
+    """Custom processor on request headers."""
+    decoded = None
+    if check_request_header(headers, "header-rs"):
```
Comment (rshriram) [inline callouts/python/extproc/example/jwt_auth/service_callout_example.py:75]: per comments above, I think we should rely on some common pattern for JWT .. i.e. decoding bearer tokens in Authorization: Bearer .. or some other format that is commonly used.

File: callouts/python/extproc/example/jwt_auth/service_callout_example.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      decoded = validate_jwt_authorization(self.public_key, context, "RS256")
+    if check_request_header(headers, "header-hmac"):
+      decoded = validate_jwt_authorization('service-extensions', context, "HS256")
```
Comment (rshriram) [inline callouts/python/extproc/example/jwt_auth/service_callout_example.py:78]: we dont need hmac. Its sufficient to do one which is RS256.

**Comment by pweiber** [comment]
> @pweiber we can close #65 in favor of #70 . Suggested some fixes

@rshriram let me know if I need to adjust something else.
Also for some reason in the pipeline the token is being invalid, locally it is working fine, also validated it in the jwt.io, is there any config I need to set for that?

**Review by jstraceski — DISMISSED** [review]

**Comment by jstraceski** [comment]
The tests seem to be failing still. Are they passing locally?

**Comment by pweiber** [comment]
@jstraceski I believe it should be fine now, I had to generate it on the fly and for the iat and exp just validate the pattern, the values were not matching by the time of the assert, let me know what you think

**Review by rshriram — APPROVED** [review]

**Pull request merged**